torch/_decomp/decompositions.py

```diff
@@ -1983,6 +1983,131 @@ def get_scale_value(scales, idx):
+    # so we cannot redispatch to aten.upsample_nearest2d.default here
+    return upsample_nearest2d(input, osize, scale_h, scale_w)
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest3d.vec)
+@torch.ops.aten.upsample_nearest3d.vec.py_impl(DispatchKey.CompositeImplicitAutograd)
+@torch.ops.aten.upsample_nearest3d.vec.py_impl(DispatchKey.Autograd)
+def upsample_nearest3d_vec(input, output_size, scale_factors):
+    osize = upsample_compute_output_size(input.size(), output_size, scale_factors)
+    scale_d = get_scale_value(scale_factors, 0)
+    scale_h = get_scale_value(scale_factors, 1)
+    scale_w = get_scale_value(scale_factors, 2)
+
+    # NB: osize could be a list of float when scale_factors is float
+    # so we cannot redispatch to aten.upsample_nearest3d.default here
+    return upsample_nearest3d(input, osize, scale_d, scale_h, scale_w)
+
+
+def _compute_upsample_nearest_indices(input, output_size):
+    # For each dim in output_size, compute the set of input indices used
+    # to produce the upsampled output.
+    indices = []
+    num_spatial_dims = len(output_size)
+    for d in range(num_spatial_dims):
+        # Math matches aten/src/ATen/native/cpu/UpSampleKernel.cpp
+        # Indices are computed as following:
+        # scale = isize / osize
+        # input_index = floor(output_index * scale)
+        # Same as OpenCV INTER_NEAREST
+        osize = sym_float(output_size[d])
+        output_indices = torch.arange(sym_int(osize), dtype=input.dtype, device=input.device)
+        isize = sym_float(input.shape[-num_spatial_dims + d])
+        scale = isize / osize
+        input_indices = torch.floor(output_indices * scale).to(torch.int64)
+        for _ in range(num_spatial_dims - 1 - d):
+            input_indices = input_indices.unsqueeze(-1)
+        indices.append(input_indices)
+    return tuple(indices)
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest1d.default)
+@torch.ops.aten.upsample_nearest1d.default.py_impl(DispatchKey.Autograd)
+@pw_cast_for_opmath
+def upsample_nearest1d(
+    input: Tensor,
+    output_size: List[Union[int, float]],
+    scales: Optional[float] = None,
+) -> Tensor:
+    l_indices, = _compute_upsample_nearest_indices(input, output_size)
+    result = input[:, :, l_indices]
+    return result
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest2d.default)
+@torch.ops.aten.upsample_nearest2d.default.py_impl(DispatchKey.Autograd)
+@pw_cast_for_opmath
+def upsample_nearest2d(
+    input: Tensor,
+    output_size: List[Union[int, float]],
+    scales_h: Optional[float] = None,
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    h_indices, w_indices = _compute_upsample_nearest_indices(input, output_size)
+    result = input[:, :, h_indices, w_indices]
+
+    # convert output to correct memory format, if necessary
+    memory_format = utils.suggest_memory_format(input)
+
+    # following "heuristic: only use channels_last path when it's faster than the contiguous path"
```

Comment:
I stole this heuristic from the `upsample_bilinear2d` decomp below, but I'm not sure where it came from or whether it's correct here.

Reply:
@jbschlosser can you detail a bit more this heuristic ?

I could find a reference in cuda code, but I could not find there where channels first memory format is set explicitly:

https://github.com/pytorch/pytorch/blob/0acfe2ce096b874303f1325c5188c457d18d828d/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L270-L331

Here is a repro code that can show memory format inconsistency vs aten:
```python
import torch

device = "cuda"


x = torch.ones(2, 3, 32, 32, dtype=torch.float32, device=device).contiguous(memory_format=torch.channels_last)
print("Input:", x.shape, x.stride(), x.device, x.dtype)


expected_output = torch._C._nn.upsample_nearest2d(x, [12, 12])
print(
    "Output, Aten:",
    expected_output.shape,
    expected_output.is_contiguous(),
    expected_output.is_contiguous(memory_format=torch.channels_last),
)

fn = torch.compile(torch.ops.aten.upsample_nearest2d.default)
decomp_output = fn(x, [12, 12])
print(
    "Output, Decomposition:",
    decomp_output.shape,
    decomp_output.is_contiguous(),
    decomp_output.is_contiguous(memory_format=torch.channels_last),
)
```

```
Input: torch.Size([2, 3, 32, 32]) (3072, 1, 96, 3) cuda:0 torch.float32
Output, Aten: torch.Size([2, 3, 12, 12]) False True
Output, Decomposition: torch.Size([2, 3, 12, 12]) True False
```